include/vcpkg/base/messages.h

```diff
@@ -0,0 +1,202 @@
+        error = 0xC,    // FOREGROUND_RED | FOREGROUND_INTENSITY
+        warning = 0xE,  // FOREGROUND_GREEN | FOREGROUND_RED | FOREGROUND_INTENSITY
+    };
+#else
+    enum class Color : char
+    {
+        none = 0,
+        success = '2', // [with 9] bright green
+        error = '1',   // [with 9] bright red
+        warning = '3', // [with 9] bright yellow
+    };
+#endif
+}
+
+namespace vcpkg::msg
+{
+    struct LocalizedString;
+
+    namespace detail
+    {
+        template<class Tag, class Type>
+        struct MessageArgument
+        {
+            const Type* parameter; // always valid
+        };
+
+        template<class... Tags>
+        struct MessageCheckFormatArgs
+        {
+            static constexpr void check_format_args(const Tags&...) noexcept { }
+        };
+
+        LocalizedString internal_vformat(::size_t index, fmt::format_args args);
+
+        template<class... Args>
+        MessageCheckFormatArgs<Args...> make_message_check_format_args(const Args&... args);
+
+        ::size_t startup_register_message(StringLiteral name, StringLiteral format_string, StringLiteral comment);
+
+        ::size_t number_of_messages();
+
+        // REQUIRES: index < last_message_index()
+        StringView get_format_string(::size_t index);
+        // REQUIRES: index < last_message_index()
+        StringView get_message_name(::size_t index);
+        // REQUIRES: index < last_message_index()
+        StringView get_default_format_string(::size_t index);
+        // REQUIRES: index < last_message_index()
+        StringView get_localization_comment(::size_t index);
+    }
+
+    // load from "locale_base/${language}.json"
+    void threadunsafe_initialize_context(const Filesystem& fs, StringView language, const Path& locale_base);
+    // load from the json object
+    void threadunsafe_initialize_context(const Json::Object& message_map);
+    // initialize without any localized messages (use default messages only)
+    void threadunsafe_initialize_context();
+
+    struct LocalizedString
+    {
+        LocalizedString() = default;
+        operator StringView() const { return m_data; }
+        const std::string& data() const { return m_data; }
+
+        static LocalizedString from_string_unchecked(std::string&& s)
+        {
+            LocalizedString res;
+            res.m_data = std::move(s);
+            return res;
+        }
+
+        LocalizedString& append(const LocalizedString& s);
+
+    private:
+        std::string m_data;
+        // to avoid lock-in on LocalizedString, these are free functions
+        // this allows us to convert to `std::string` in the future without changing All The Code
+        friend LocalizedString& append_newline(LocalizedString&);
+        friend LocalizedString&& append_newline(LocalizedString&& self) { return std::move(append_newline(self)); }
+        friend LocalizedString& appendnl(LocalizedString& self, const LocalizedString& to_append)
```

Comment:
I would prefer to not add different spellings of the same op; can we either make them all "append_newline" or all "appendnl"?